src/qibojit/custom_operators/backends.py

```diff
@@ -384,3 +383,173 @@ def swap_pieces(self, piece0, piece1, new_global, nlocal):
+        self.cuquantum = cuquantum 
+        self.cusv = cusv
+        self.name = "cuquantum"
+
+    def one_qubit_base(self, state, nqubits, target, kernel, qubits=None, gate=None):
+        handle = self.cusv.create()
+        ncontrols = len(qubits) - 1 if qubits is not None else 0
+        controls = self.np.asarray(self.np.arange(ncontrols), dtype=self.np.int32)
+        # TODO: improve this dummy conversion from little to big endian for 2 qubits
+        if nqubits == 2:
+            state[[1,2]] = state[[2,1]]
+
+        state = self.cast(state, self.np.complex64)
+        ntarget = 1
+        target = self.np.asarray([target], dtype=self.np.int32)
+        adjoint = 0
+        # implement x, y, z and z_pow kernel since cuquantum only has apply_matrix
+        # TODO: find a better way to implement this
+        if gate is None:
+            gate = self.cp.zeros((2, 2))
+            if kernel == "apply_x":
+                gate[0, 1], gate[1, 0] = 1, 1
+            elif kernel == "apply_y":
+                gate[0, 1], gate[1, 0] = -1j, 1j
+            elif kernel == "apply_z":
+                gate[0, 0], gate[1, 1] = 1, -1
+        gate = self.cast(gate, self.np.complex64)
+        if isinstance(gate, self.cp.ndarray):
+            gate_ptr = gate.data.ptr
+        elif isinstance(gate, self.np.ndarray):
+            gate_ptr = gate.ctypes.data
+        else:
+            raise ValueError
+
+        args1 = (handle,
+                 self.cuquantum.cudaDataType.CUDA_C_32F,
+                 nqubits,
+                 gate_ptr,
+                 self.cuquantum.cudaDataType.CUDA_C_32F,
+                 self.cusv.MatrixLayout.ROW,
+                 adjoint,
+                 ntarget,
+                 ncontrols,
+                 self.cuquantum.ComputeType.COMPUTE_32F
+                 )
+        workspaceSize = self.cusv.apply_matrix_buffer_size(*args1)
+
+        # check the size of external workspace
+        if workspaceSize > 0:
+            workspace = self.cp.cuda.memory.alloc(workspaceSize)
+            workspace_ptr = workspace.ptr
+        else:
+            workspace_ptr = 0
+
+        args2 = (handle,
+                 state.data.ptr,
+                 self.cuquantum.cudaDataType.CUDA_C_32F,
+                 nqubits,
+                 gate_ptr,
+                 self.cuquantum.cudaDataType.CUDA_C_32F,
+                 self.cusv.MatrixLayout.ROW,
+                 adjoint,
+                 target.ctypes.data,
+                 ntarget,
+                 controls.ctypes.data,
+                 ncontrols,
+                 0,
+                 self.cuquantum.ComputeType.COMPUTE_32F,
+                 workspace_ptr,
+                 workspaceSize
+                 )
+
+        self.cusv.apply_matrix(*args2)
+        # TODO: improve this dummy conversion from little to big endian for 2 qubits
+        if len(qubits) == 2:
+            state[[1,2]] = state[[2,1]]
```

Comment:
I am not sure if I understand this correctly but the general case of this would probably look like:
```Python
original_shape = state.shape
state = cp.reshape(state, nqubits * (2,))
state = cp.transpose(state, range(nqubits - 1, -1, -1))
state = cp.reshape(state, original_shape)
```
If yes, then keep in mind that this is a very costly operation which we would probably want to avoid if we want this backend to have competitive performance. Ideally the difference regarding the most significant qubit should be taken into account in the indices we pass in the kernel, if this is possible.